### ISSUE TITLE
Properties are not skipped

### ISSUE THREAD

**Issue by tavolate** [modelmapper/modelmapper#217]
In the last release 0.7.8 the same code working on v0.7.7 doesn't work.

If I try to map a source object into a destination object with some not existing properties in source object, in the destination object the value is set to null even if I configurated to skip these properties

**Comment by chhsiao90** [comment]
Thanks for opening the report.
I think it related to b17492f6ee7e1bf61e5b4b1a48a89b84a61ac67f and #197 
Can you provide your use case?

**Comment by tavolate** [comment]
private void init() {        
       ModelMapper mModelMapper = new ModelMapper();

        Condition<Object, Object> lNotNull = new Condition<Object, Object>() {
            public boolean applies(MappingContext<Object, Object> pContext) {
                return pContext.getSource() != null;
            }
        };
       mModelMapper.getConfiguration().setPropertyCondition(lNotNull);
       mModelMapper.addMappings(new UserDtoToActiveUserEntity());
    }

    private class UserDtoToActiveUserEntity extends PropertyMap<UserDto, ActiveUserEntity> {
        @Override
        protected void configure() {
            skip(destination.getId());
            skip(destination.getUuid());
        }
    }

Id and Uuid is not present in UserDto and when I map the userDto to the ActivieUserEntity the id and uuid are set to null

**Comment by chhsiao90** [comment]
Thanks for the sample code.
You said destination's id and uuid is not skipped, which source properties is mapped to destination?
I had tried following sample code, but still cannot reproduce your issue.
Please correct me if I'm doing wrong.
```java
@Test
public class GC217 extends AbstractTest {
  static class UserDto {
    private String userId;

    public String getUserId() {
      return userId;
    }

    public void setUserId(String userId) {
      this.userId = userId;
    }
  }

  static class ActiveUserEntity {
    private String id;
    private String uuid;

    public String getId() {
      return id;
    }

    public void setId(String id) {
      this.id = id;
    }

    public String getUuid() {
      return uuid;
    }

    public void setUuid(String uuid) {
      this.uuid = uuid;
    }
  }

  public void shouldMap() {
    modelMapper.getConfiguration().setPropertyCondition(new Condition<Object, Object>() {
      public boolean applies(MappingContext<Object, Object> pContext) {
        return pContext.getSource() != null;
      }
    });
    modelMapper.addMappings(new PropertyMap<UserDto, ActiveUserEntity>() {
      @Override
      protected void configure() {
        skip(destination.getId());
        skip(destination.getUuid());
      }
    });

    ActiveUserEntity entity = modelMapper.map(new UserDto(), ActiveUserEntity.class);
    assertNotNull(entity);
    assertNull(entity.getId());
    assertNull(entity.getUuid());
  }
}
```

**Comment by tavolate** [comment]
This test fail with the 0.7.8
    
    public class ModelMapperTest {
     static class Source {
        private String id;

        public String getId() {
            return id;
        }

        public void setId(String pId) {
            id = pId;
        }
    }

    static class Destination {
        private String id;
        private String uuid;

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public String getUuid() {
            return uuid;
        }

        public void setUuid(String uuid) {
            this.uuid = uuid;
        }
    }

    @Test
    public void shouldMap() {
        ModelMapper lModelMapper = new ModelMapper();
        lModelMapper.getConfiguration().setPropertyCondition(new Condition<Object, Object>() {
            public boolean applies(MappingContext<Object, Object> pContext) {
                return pContext.getSource() != null;
            }
        });
        lModelMapper.addMappings(new PropertyMap<Source, Destination>() {
            @Override
            protected void configure() {
                skip(destination.getId());
                skip(destination.getUuid());
            }
        });

        Source lSrc = new Source();
        lSrc.setId("5555");

        Destination lDest = new Destination();
        lDest.setId("1234");
        lDest.setUuid("AABB");

        lModelMapper.map(lSrc, lDest);
        Assert.assertNotNull(lDest);
        Assert.assertNotNull(lDest.getId());
        Assert.assertNotNull(lDest.getUuid());
     }
    }

**Comment by chhsiao90** [comment]
OK, I had misunderstand by this..
> If I try to map a source object into a destination object with some not existing properties in source object

I will take a look!

**Comment by chhsiao90** [comment]
@tavolate -
Thanks for your reproducible code and I had figured out the reason.
I would like to know is there any reason that you need to defined a global property condition?
```
    lModelMapper.getConfiguration().setPropertyCondition(new Condition<Object, Object>() {
        public boolean applies(MappingContext<Object, Object> pContext) {
            return pContext.getSource() != null;
        }
    });
```

It will turn the PropertyMap map/skip into this kind of logic
when(condition).skip(...)
-> if condition true skip the property mapping
-> if condition false, it won't skip the destination mapping, so it will invoke the property mapping even the source property is null

I will prefer ModelMapper as is.
But I still would like to know what's your use case, and maybe we can figure a best solution!

**Comment by tavolate** [comment]
I use the property condition globally just for simplicity. I don't' want to repeat the condition for every mapping

**Comment by tavolate** [comment]
The only problem with your solution is I have to know before if a property is null. My common logic is "if some property from the source object is null, these properties must be skipped"

**Comment by chhsiao90** [comment]
> The only problem with your solution is I have to know before if a property is null.

I think it's not true.
You only need this when you want skip something.
For above example, if you want to map id when it is not null, you can do like this.
```java
    modelMapper.addMappings(new PropertyMap<UserDto, ActiveUserEntity>() {
      @Override
      protected void configure() {
        map(source.getId(), destination.getId());
        when(alwaysTrue).skip(null, destination.getUuid());
      }
    });
```

**Comment by tavolate** [comment]
with your solution, I have to write

`when(alwaysTrue).skip(null, destination.getUuid());`

for every property in the destination object if I want to skip the property when the source property is null

**Comment by chhsiao90** [comment]
Yeah...
from ```skip(destination.getUuid())``` to ```when(alwaysTrue).skip(null, destination.getUuid())``` or ```when(alwaysTrue).skip().setUuid(null)```

Or you should prevent from using skip if you config ```nonNull``` property condition into your modelMapper.
Following code I think it works too.
```
   modelMapper.addMappings(new PropertyMap<UserDto, ActiveUserEntity>() {
      @Override
      protected void configure() {
        map(source.getId(), destination.getId());
        map.setUuid(null);
      }
    });
```

**Comment by tavolate** [comment]
Excuse me If I am pedantic but with the old solution, with a global definition, 

      Condition<Object, Object> lNotNull = new Condition<Object, Object>() {
          public boolean applies(MappingContext<Object, Object> pContext) {
              return pContext.getSource() != null;
          }
       };

I could manage every mapping even if it isn't explicitly defined. I have about 20 objects with 10 properties for each one and I wouldn't want to write 200 times

`when(alwaysTrue).skip().setUuid(null)`

**Comment by chhsiao90** [comment]
Or maybe you should use Strict [Matching Strategy](http://modelmapper.org/user-manual/configuration/#matching-strategies) and just ignore the properties not exists in source?

**Comment by tavolate** [comment]
So, Do we have any chance to use a global definition like we were used to using with v0.7.7?

**Comment by chhsiao90** [comment]
The logic of #197 is what we expect and the detail had described at [comment](https://github.com/modelmapper/modelmapper/issues/217#issuecomment-294284006).
I don't plan to revert the change because I think it is right.

Did you already have ```skip(destination.getXxx())``` **200** times in your current code base?
So you need to change all these code if you want to upgrade to v0.7.8?

I think the only way to solve is [issuecomment-294734680](https://github.com/modelmapper/modelmapper/issues/217#issuecomment-294734680) or [issuecomment-294628142](https://github.com/modelmapper/modelmapper/issues/217#issuecomment-294628142) in current version of ModelMapper.

But I will be more interested at why there need so much skip, or need an global condition to apply "if some property from the source object is null, these properties must be skipped".
I think there might be something that ModelMapper can do it better.
One of my idea is to had a global config flag for ```skipIfNull``` or something like this.
So you can just set ```modelMapper.getConfiguration().setSkipIfNull(true);```
But you still need to removed the 200 times ```skip(destination.getUuid())``` if we had the flag...

**Comment by tavolate** [comment]
We have an open API and our customers are able to valorize or not some properties of the DTO. Whenever the customer does not valorize a property, its value is null. When the property is null I want to modelmapper skips the mapping into the destination property (same name)

A something like this `modelMapper.getConfiguration().setSkipIfNull(true);` will be appreciated.

I don't have 200 times `skip(destination.getUuid())` because with the old version, it was not needed.

**Comment by chhsiao90** [comment]
OK, thanks!
I'm glad that the skipIfNull might solve your issue and I will had that in next release.

But I'm still a little confused about you use case.

You had said
>  don't have 200 times skip(destination.getUuid()) because with the old version, it was not needed.

and said
> I could manage every mapping even if it isn't explicitly defined. I have about 20 objects with 10 properties for each one and I wouldn't want to write 200 times

The issue is about you had doing following actions in the same time
- When you configured a  global property condition
- You had used skip in some of your PropertyMap

And you need modify all the usage of skip to when(alwaysTrue).skip().setUuid() when you match both conditions above.

Let me said it clearly,
- if you **don't have any skip** in your old code, you don't need ANY change to upgrade to v0.7.8
- If you **had just few skip** in your old code, you just need to change that **few** skip.

So I don't know why you need 200 times change if you don't have 200 times skip in your old code.

**Comment by tavolate** [comment]
Sorry, you are right. I thought to upgrade to v0.7.8 I had to add 200 times skip.

Thank you!

**Comment by chhsiao90** [comment]
I'm not challenge your use case, just still feel a little confused.
I will still apply ```setSkipNull``` feature.
So I will re-open this and close until we apply the feature!

**Comment by chhsiao90** [comment]
@tavolate - we had #255 to be released with v1.1.0. Please have a try and see it can fix your problem, thanks!
